### PR TITLE
implement ability to require deployment reasons per environment and notify from workspace conf

### PR DESCRIPTION
**Pull request by n-g**
Fixes NSL-2788
Fixes NSL-2787

### CODE REVIEW

**Comment by n-g** [comment]
Back to WIP:
I want to add the slack notification settings from workspace and secrets so that we see the full config picture.

**Comment by n-g** [comment]
Fixes NSL-2787
Fixes NSL-2788

File: internal/cli/cmd/deploy.go [hunk 1]
```diff
@@ -71,12 +72,17 @@ func NewDeployCmd() *cobra.Command {
...
 			flags.BoolVar(&uploadToRegistry, "upload_to_registry", false, "If set, uploads the deploy plan to the cluster registry, instead of applying it.")
 			flags.StringVar(&deployOpts.outputPath, "output_to", "", "If set, a machine-readable output is emitted after successful deployment.")
+			flags.StringVar(&deployOpts.manualReason, "reason", "", "Why was this deployment triggered.")
 		}).
 		With(
 			fncobra.ParseEnv(&env),
 			fncobra.ParseLocations(&locs, &env, fncobra.ParseLocationsOpts{ReturnAllIfNoneSpecified: true}),
 			planningargs.ParseServers(&servers, &env, &locs)).
 		Do(func(ctx context.Context) error {
+			if serializePath == "" && getDeployReason(deployOpts) == "" && deploy.RequireReason(env.Configuration()) {
```
Comment (hugosantos) [inline internal/cli/cmd/deploy.go:82]: I'd check for `deploy.RequireReason` first for readability purposes.

File: internal/cli/cmd/deploy.go [hunk 2]
```diff
@@ -286,3 +293,15 @@ func uploadPlanTo(ctx context.Context, target compute.Computable[oci.RepositoryW
 	fmt.Fprintf(console.Stdout(ctx), "Pushed plan to %s\n", resultImageID.RepoAndDigest())
 	return nil
 }
+
+func getDeployReason(opts deployOpts) string {
```
Comment (hugosantos) [inline internal/cli/cmd/deploy.go:297]: s/getDeployReason/deployReason/

File: internal/cli/cmd/deployplan.go [hunk 3]
```diff
@@ -55,6 +57,10 @@ func NewDeployPlanCmd() *cobra.Command {
 			return err
 		}
 
+		if getDeployReason(opts) == "" && deploy.RequireReason(config) {
```
Comment (hugosantos) [inline internal/cli/cmd/deployplan.go:60]: Same here, deploy.RequireReason() first, as that's what's triggering the condition.

File: orchestration/migration.go [hunk 4]
```diff
@@ -36,26 +35,27 @@ func ExecuteOpts() execution.ExecuteOpts {
...
 	if !DeployWithOrchestrator {
 		if !wait {
 			return fnerrors.BadInputError("waiting is mandatory without the orchestrator")
 		}
 
 		observeError := func(context.Context, error) {}
-		if DeployUpdateSlackChannel != "" {
-			if SlackToken == "" {
+		if token, channel, err := getSlackTokenAndChannel(ctx, env); err != nil {
+		} else if channel != "" {
```
Comment (hugosantos) [inline orchestration/migration.go:46]: missing body?

File: orchestration/slack.go [hunk 5]
```diff
@@ -5,21 +5,61 @@
...
+	"namespacelabs.dev/foundation/framework/secrets"
+	"namespacelabs.dev/foundation/framework/secrets/combined"
+	"namespacelabs.dev/foundation/internal/parsing"
+	"namespacelabs.dev/foundation/internal/planning/deploy"
 	"namespacelabs.dev/foundation/schema"
+	"namespacelabs.dev/foundation/std/cfg"
 )
 
-func renderSlackMessage(plan *schema.DeployPlan, start, end time.Time, err error) []slack.Block {
+func getSlackTokenAndChannel(ctx context.Context, env cfg.Context) (string, string, error) {
```
Comment (hugosantos) [inline orchestration/slack.go:24]: I'd go with `resolveSlack...`

**Review by hugosantos — APPROVED** [review]

File: orchestration/migration.go (hunk 4 above)
Comment (n-g) [reply orchestration/migration.go:46]: thanks!

**Pull request merged**